### ISSUE TITLE
💡 RFC: Using frontmatter values in script tags

### ISSUE THREAD

**Issue by Tc-001** [snowpackjs/astro#1107]
### Background & Motivation

While Astro is good for server-side prerendering, there may be many cases one might wish to use a server-side variable in client side, inline javascript. 

### Possible solutions
- `use:values` attribute could be used to add user-selected variables as constants to the script.
- Non-js-clashing alternate syntax for dynamic values (ex. `const foo = %{bar}`)

### Risks, downsides, and/or tradeoffs
- If not properly sanitised or escaped, all approaces could be prone to XSS, especially if building from an open/compromised database or API

### Open Questions
- If using dynamic values, is there a way to keep `${foo}`, `%{foo}`, `%foo%` or any other syntax both usable and not trigger false positives in the compiler (ex. `const allowed_chars = "abcdxyz,./+-*/%{}[]"` with `%{}` would be wrongly replaced if using something like a regex replacement)
- How will the data be serialised? Something as simple as `JSON.stringify`, or full on byte64?
- (both cases) Will it be intuitive to the user what the script does to make it work?
- Will this kind of insertion clash with typescript?

### For `use:values` 
@FredKSchott suggested inserting constants at the script top
Input:
```astro
---
const foo = 42;
---
<script use:values={{foo}}>
  console.log('in the browser, foo is', foo);
</script>
```

Output:
All values are stringified and turned into const declarations.
```html
<!-- in the browser: -->
<script>
  const foo = 42;
  console.log('in the browser, foo is', foo);
</script>
```
### For alternate dynamic values
A simple JS parser of some kind could be used to look for the specific syntax the dynamic value uses, but there could be some edge cases that would need to be discussed/accounted for

### Help make it happen!

- [ ] I am willing to submit a PR to implement this change.
- [X] I am willing to submit a PR to implement this change, but would need some guidance.
- [ ] I am not willing to submit a PR to implement this change.

**Comment by jonathantneal** [comment]
Love it. In [terser](https://terser.org/docs/api-reference#conditional-compilation) and [esbuild](https://esbuild.github.io/api/#define), I notice this kind of functionality is termed "_define_".

Therefore, for your consideration:

```astro
---
const answer = 42;
---
<script define:const={{ answer }}>
  console.log('the answer is', answer);
</script>
```

... and then maybe someday ...

```astro
<style define:var={{ answer }}>
  body {
    margin: calc(1px * var(--answer));
  }
</style>
```

**Comment by RafidMuhymin** [comment]
## Detailed Design
### For %{} syntax

You'll be able to access frontmatter values using the %{} syntax.

Input:
```js
---
    const foo = "foo"
---

<script>
    console.log(%{foo})
</script> 
```

The compiler will consider the content of the script tag as plain text. Then it'll insert the frontmatter variables like JSX.

Output:
 ```js
<script>
    console.log("foo")
</script> 
```

## Downsides
Adding a new non-native operator may cause clash with the language, mostly with string values.

## Possible Alternative Syntax
Just use the native ${} syntax. But to implement this, the Astro compiler will have to check if a local or global variable  or is lo

**Comment by RafidMuhymin** [comment]
## Detailed Design
### For %{} syntax

You'll be able to access frontmatter values using a new syntax, for example %{}.

Input:
```js
---
    const foo = "foo"
---

<script>
    console.log(%{foo})
</script> 
```

The compiler will consider the content of the script tag as plain text. Then it'll insert the frontmatter variables like JSX.

Output:
 ```js
<script>
    console.log("foo")
</script> 
```

## Downsides
Adding a new non-native operator may cause clash with the language, mostly with string values.

## Possible Alternative Syntax
Just use the native ${} syntax. But to implement this, the Astro compiler will have to check if a local or global variable is declared with that name in every script.

**Comment by FredKSchott** [comment]
This has come up a few times in Discord now! I think we should have official support for this via a new directive. Happy with `use:values`, `define:values`, `use:data`, `define:data`, and most variations. A slight, completely subjective and personal preference -1 for "vars" just because it always reminds me of `var` which I feel has a slight... odor to it in most developers eyes :)

Also for anyone reading this looking for a solution today, we have a workaround: Use `<script data-foo={foo}>` and then read from the `data-foo` attribute in the script itself.

**Comment by natemoo-re** [comment]
Love this idea! I would really prefer a directive rather than a special escaping syntax. A directive would keep bloat out of the compiler and simplify syntax highlighting logic.

Also +1 on any sort of variation. I do like `define` over `use`. `define:vars` is kinda nice because we still use "variable" to refer to `const`/`let` values and colloquially to `--custom-properties`. So there's a nice symmetry there.

**Comment by FredKSchott** [comment]
If consensus is aligning around `define:vars` then I'm +1

**Comment by FredKSchott** [comment]
I'll throw out one other idea: `<script define:foo={foo} define:bar={`hello ${name}`}>`

**Comment by Tc-001** [comment]
+1 to that, especially if you can do `define:{foo, bar, baz}` or similar

**Comment by FredKSchott** [comment]
Just calling out: this probably will not be allowed with the `hoist` ability being implemented in #1077 (since every instance of every component would have a different value, but a hoisted script only exists once).

**Comment by matthewp** [comment]
This didn't make the cut for the call this week but I think we're getting close to being aligned on this so it would be nice to talk about next week.

**Comment by jasikpark** [comment]
An interesting solution to this is
```astro
---
const { interpolatedVar } = Astro.props;
const scriptTag = `<script type="module">console.log(${interpolatedVar});</script>`;
---
{scriptTag}
```
though then you wouldn't be able to use tagged templates in the script tag

**Comment by okikio** [comment]
Would this work for [JSON/LD](https://developers.google.com/search/docs/advanced/structured-data/intro-structured-data) SEO files?

**Comment by jasikpark** [comment]
> Would this work for [JSON/LD](https://developers.google.com/search/docs/advanced/structured-data/intro-structured-data) SEO files?

or should those files be special cased in the compiler as not scripts but data tags?

**Comment by okikio** [comment]
> > Would this work for [JSON/LD](https://developers.google.com/search/docs/advanced/structured-data/intro-structured-data) SEO files?
> 
> or should those files be special cased in the compiler as not scripts but data tags?

That would also work

**Comment by radiosilence** [comment]
I wonder if it would be simpler to do something like

```
---

import { injectValues } from 'astro';

injectValues({myVar: 'potato' });

---

<script>
  console.log(myVar);
</script>
```

And then somewhere in the output:

```
globals.myVar = 'potato';
```

This would work with serverside rendering and would require no transformation.

**Comment by radiosilence** [comment]
For now I am doing this

```astro
---

const gaId = 'A-123123';
---

<html>
  <meta id="ga-id" content={gaId} />

...

<script>
   console.log(document.getElementById, 'ga-id').content
</script>
```

**Comment by FredKSchott** [comment]
just FYI I think @natemoo-re has an implementation of this in the new compiler, which should be put into a PR next week!

**Comment by sacarvy** [comment]
For Now, What I am doing,
```astro
---
const { uid } = Astro.props // unique ID
---
<div id={uid} >

</div>
{`<script>let element${uid} = document.getElementById('${uid}'); console.log(element${uid})</script>`}
```

**Comment by jasikpark** [comment]
This is being added in the new compiler:

- [`<script hoist define:vars={{builtAtFormatted}}>`](https://astro-repl-git-feat-update-compiler-pikapkg.vercel.app/play#eyJpbm1lbW9yeTovL21vZGVsL3NyYy9wYWdlcy9QYWdlLmFzdHJvIjoiLS0tXG4vLyBVc2luZyBkZWZpbmU6dmFycyB3aXRoIGEgc2NyaXB0XG5pbXBvcnQge2Zvcm1hdH0gZnJvbSAnZGF0ZS1mbnMnOyBcbmNvbnN0IGJ1aWx0QXQ6IERhdGUgPSBuZXcgRGF0ZSgpO1xuY29uc3QgYnVpbHRBdEZvcm1hdHRlZCA9IGZvcm1hdChidWlsdEF0LCAnTU1NTSBkZCwgeXl5eSAtLSBIOm1tOnNzLlNTUycpO1xuLS0tXG48aHRtbCBsYW5nPVwiZW5cIj5cbiAgPGhlYWQ+XG4gICAgPG1ldGEgY2hhcnNldD1cIlVURi04XCI+XG4gICAgPHRpdGxlPkFzdHJvIFBsYXlncm91bmQ8L3RpdGxlPlxuICAgIDxzdHlsZT5cbiAgICAgIGhlYWRlciB7XG4gICAgICAgIGRpc3BsYXk6IGZsZXg7XG4gICAgICAgIGZsZXgtZGlyZWN0aW9uOiBjb2x1bW47XG4gICAgICAgIGFsaWduLWl0ZW1zOiBjZW50ZXI7XG4gICAgICAgIHRleHQtYWxpZ246IGNlbnRlcjtcbiAgICAgICAgbWFyZ2luLXRvcDogMTV2aDtcbiAgICAgICAgZm9udC1mYW1pbHk6IEFyaWFsO1xuICAgICAgfVxuICAgICAgLm5vdGUge1xuICAgICAgICBtYXJnaW46IDA7XG4gICAgICAgIHBhZGRpbmc6IDFyZW07XG4gICAgICAgIGJvcmRlci1yYWRpdXM6IDhweDtcbiAgICAgICAgYmFja2dyb3VuZDogI0U0RTVFNjtcbiAgICAgICAgYm9yZGVyOiAxcHggc29saWQgI0JCQjtcbiAgICAgIH1cbiAgICA8L3N0eWxlPlxuICA8L2hlYWQ+XG4gIDxib2R5PlxuICAgIDxoZWFkZXI+XG4gICAgICA8aW1nIHdpZHRoPVwiNjBcIiBoZWlnaHQ9XCI4MFwiIHNyYz1cImh0dHBzOi8vYmVzdG9manMub3JnL2xvZ29zL2FzdHJvLnN2Z1wiIGFsdD1cIkFzdHJvIGxvZ29cIj5cbiAgICAgIDxoMT5IZWxsbywgQXN0cm8hPC9oMT5cbiAgICAgIDxwIGNsYXNzPVwibm90ZVwiPlxuICAgICAgICA8c3Ryb25nPlJFTkRFUkVEIEFUOjwvc3Ryb25nPjxici8+XG4gICAgICAgIHtidWlsdEF0Rm9ybWF0dGVkfVxuICAgICAgPC9wPlxuICAgIDwvaGVhZGVyPlxuICAgIDwhLS0gVGhpcyB3aWxsIGxvZyBvdXQgaW4gdGhlIGNvbnNvbGUhIC0tPlxuICAgIDxzY3JpcHQgaG9pc3QgZGVmaW5lOnZhcnM9e3sgYnVpbHRBdEZvcm1hdHRlZCB9fT5jb25zb2xlLmxvZyhidWlsdEF0Rm9ybWF0dGVkKTwvc2NyaXB0PlxuICA8L2JvZHk+XG48L2h0bWw+XG4ifQ==)
- [`<script define:vars={{builtAtFormatted}}>`](https://astro-repl-git-feat-update-compiler-pikapkg.vercel.app/play#eyJpbm1lbW9yeTovL21vZGVsL3NyYy9wYWdlcy9QYWdlLmFzdHJvIjoiLS0tXG4vLyBVc2luZyBkZWZpbmU6dmFycyB3aXRoIGEgc2NyaXB0XG5pbXBvcnQge2Zvcm1hdH0gZnJvbSAnZGF0ZS1mbnMnOyBcbmNvbnN0IGJ1aWx0QXQ6IERhdGUgPSBuZXcgRGF0ZSgpO1xuY29uc3QgYnVpbHRBdEZvcm1hdHRlZCA9IGZvcm1hdChidWlsdEF0LCAnTU1NTSBkZCwgeXl5eSAtLSBIOm1tOnNzLlNTUycpO1xuLS0tXG48aHRtbCBsYW5nPVwiZW5cIj5cbiAgPGhlYWQ+XG4gICAgPG1ldGEgY2hhcnNldD1cIlVURi04XCI+XG4gICAgPHRpdGxlPkFzdHJvIFBsYXlncm91bmQ8L3RpdGxlPlxuICAgIDxzdHlsZT5cbiAgICAgIGhlYWRlciB7XG4gICAgICAgIGRpc3BsYXk6IGZsZXg7XG4gICAgICAgIGZsZXgtZGlyZWN0aW9uOiBjb2x1bW47XG4gICAgICAgIGFsaWduLWl0ZW1zOiBjZW50ZXI7XG4gICAgICAgIHRleHQtYWxpZ246IGNlbnRlcjtcbiAgICAgICAgbWFyZ2luLXRvcDogMTV2aDtcbiAgICAgICAgZm9udC1mYW1pbHk6IEFyaWFsO1xuICAgICAgfVxuICAgICAgLm5vdGUge1xuICAgICAgICBtYXJnaW46IDA7XG4gICAgICAgIHBhZGRpbmc6IDFyZW07XG4gICAgICAgIGJvcmRlci1yYWRpdXM6IDhweDtcbiAgICAgICAgYmFja2dyb3VuZDogI0U0RTVFNjtcbiAgICAgICAgYm9yZGVyOiAxcHggc29saWQgI0JCQjtcbiAgICAgIH1cbiAgICA8L3N0eWxlPlxuICA8L2hlYWQ+XG4gIDxib2R5PlxuICAgIDxoZWFkZXI+XG4gICAgICA8aW1nIHdpZHRoPVwiNjBcIiBoZWlnaHQ9XCI4MFwiIHNyYz1cImh0dHBzOi8vYmVzdG9manMub3JnL2xvZ29zL2FzdHJvLnN2Z1wiIGFsdD1cIkFzdHJvIGxvZ29cIj5cbiAgICAgIDxoMT5IZWxsbywgQXN0cm8hPC9oMT5cbiAgICAgIDxwIGNsYXNzPVwibm90ZVwiPlxuICAgICAgICA8c3Ryb25nPlJFTkRFUkVEIEFUOjwvc3Ryb25nPjxici8+XG4gICAgICAgIHtidWlsdEF0Rm9ybWF0dGVkfVxuICAgICAgPC9wPlxuICAgIDwvaGVhZGVyPlxuICAgIDwhLS0gVGhpcyB3aWxsIGxvZyBvdXQgaW4gdGhlIGNvbnNvbGUhIC0tPlxuICAgIDxzY3JpcHQgZGVmaW5lOnZhcnM9e3sgYnVpbHRBdEZvcm1hdHRlZCB9fT5jb25zb2xlLmxvZyhidWlsdEF0Rm9ybWF0dGVkKTwvc2NyaXB0PlxuICA8L2JvZHk+XG48L2h0bWw+XG4ifQ==) - for some reason it doesn't work w/o hoist atm? 

Added to https://github.com/snowpackjs/astro/pull/1406/files in https://github.com/snowpackjs/astro-compiler-next/pull/26

**Comment by natemoo-re** [comment]
Confirming that this is handled in the `astro@next` preview!

Will put this in our milestone to be closed when that release is stable.

**Comment by natemoo-re** [comment]
Following up on my previous comment, the team has decided to close out issues that have been confirmed as fixed by [astro@0.21.0-next.0](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.0), [astro@0.21.0-next.1](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.1), or [astro@0.21.0-next.2](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.2). Our hope is that this will help the [v0.21 milestone](https://github.com/snowpackjs/astro/milestone/1) remain as actionable as possible.

To verify that this issue has been fixed, you may
- Open a new [`astro@next` project on StackBlitz](https://astro.new/minimal@next)
- Locally upgrade to the `astro@next` preview
    ```shell
    npm i astro@next
    # If using Framework renderers...
    npm i @astrojs/renderer-name@next
    ```

If you think we made a mistake, **please reopen this issue**!

If you run into any other problems with `astro@next`, please [open a new issue](https://github.com/snowpackjs/astro/issues/new?assignees=&labels=0.21&template=---01-bug-report.yml&title=%F0%9F%90%9B+BUG%3Aastro@next).
